src/odemis/acq/align/z_localization.py

```diff
@@ -0,0 +1,193 @@
+                                    x (dict) --> a, b ,c, d, w0 (floats), fit on the equation of Huang in x direction
+                                    y (dict) --> a, b ,c, d, w0 (floats), fit on the equation of Huang in y direction
+                                    feature_angle (float) angle of the ellipsoidal shapes w.r.t. the positive Y axis (anti-clockwise is positive)
+                                    upsample_factor (int) number with which the data is up sampled during the calibration
+                                    z_least_confusion (float) location in z in meters where the least confusion is present in the image (in focus)
+                                    z_calibration_range (tuple) --> (min, max) floats of the min and max z value in meters w.r.t. the z_least_confusion
+
+    :param z_stack_step_size (float): the step size in meters between two images in the z stack
+    :param fit_tol (float): factor for the precision of the Gaussian fit, range from 0 --> 1
+    :return:
+        z_position (float): determined z position of the feature in meter
+        warn_lvl (int): None = No warnings
+                        1 = from fmin_cg(scipy) max number of iterations exceeded
+                        2 = from fmin_cg(scipy) gradient and/or function calls were not changing
+                        3 = from fmin_cg(scipy) NaN result encountered
+                        4 = Outputted Z position is outside the defined maximum range, output is inaccurate
+                        5 = The Gaussian fit is not precise enough, probably because the image contains too much noise
+                        6 = The Gaussian fit found a feature to big for the current feature, the size > 85%
+    :raises:
+            ModuleNotFoundError if the module psf_extractor is not found.
+            KeyError if the calibration data is incomplete and does not include all the keys
+
+    """
+    warning = None  # Set the warning level to None, no warnings
+
+    if not psf_extractor:
+        raise ModuleNotFoundError("The psf_extractor module is not found. Cannot determine the Z position")
+
+    # Fit a Gaussian to the feature in the image
+    popt, pcov = psf_extractor.extractor.fit_gaussian_2D(image, theta=numpy.rad2deg(calibration_data["feature_angle"]), epsilon=1e-6)
+    sigma_x, sigma_y = popt[2:4]
+
+    # Ensure that for a normal distribution at least 95% falls within the set tolerance for the calculated standard
+    # deviation error (this means a confidence level of 1.96 time the standard deviation above and below the mean)
+    if (fit_tol / 0.98) * sigma_x < pcov[2] or (fit_tol / 0.98) * sigma_y < pcov[3]:
+        logging.warning("The image contains too much noise to determine the size of the feature accurately, "
+                        "the results may be inaccurate.\n"
+                        "The standard deviation error for determined feature size in x is: %s and for y is %s.",
+                        pcov[2], pcov[3])
```

Comment:
This is fine. However, I don't think it's always worth to be very verbose. Honestly, the pcov values are only for us, when a user reports a problem that the "Z localization didn't work". So you could have kept it short like `std dev = {pcov[2]}, {pcov[3]}.`  Anyway, that's fine as it is.